Sources/DistributedSystem/ChannelCompression.swift

```diff
@@ -0,0 +1,202 @@
+        nextIdx = (1 - nextIdx)
+        var buffer = buffers[idx]
+        if buffer.count < capacity {
+            buffer.deallocate()
+            buffer = UnsafeMutableRawBufferPointer.allocate(byteCount: capacity, alignment: 0)
+            buffers[idx] = buffer
+        }
+        return buffer
+    }
+
+    mutating func reset() {
+        for idx in 0..<buffers.count {
+            buffers[idx].deallocate()
+        }
+    }
+}
+
+final class ChannelCompressionInboundHandler: ChannelInboundHandler {
+    typealias InboundIn = ByteBuffer
+    typealias InboundOut = ByteBuffer
+
+    private let loggerBox: Box<Logger>
+    private var bufferManager = BufferManager()
+    private var lz4Stream = LZ4_streamDecode_t()
+
+    private var logger: Logger { loggerBox.value }
+
+    init(_ loggerBox: Box<Logger>) {
+        self.loggerBox = loggerBox
+        LZ4_setStreamDecode(&lz4Stream, nil, 0);
+    }
+
+    deinit {
+        bufferManager.reset()
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        var bufferIn = unwrapInboundIn(data)
+        _ = bufferIn.readInteger(as: UInt32.self) // skip message size
+        if let messageSize = bufferIn.readInteger(as: UInt32.self) {
+            let dstBuffer = bufferManager.getNextBuffer(capacity: Int(messageSize))
+            let decompressedBytes = bufferIn.withUnsafeReadableBytes { ptr in
+                LZ4_decompress_safe_continue(&lz4Stream,
+                                             ptr.baseAddress!.assumingMemoryBound(to: Int8.self),
+                                             dstBuffer.baseAddress!.assumingMemoryBound(to: Int8.self),
+                                             Int32(ptr.count),
+                                             Int32(dstBuffer.count))
+            }
+            if decompressedBytes == messageSize {
+                var bufferOut = ByteBufferAllocator().buffer(capacity: MemoryLayout<UInt32>.size + Int(messageSize))
+                bufferOut.writeInteger(messageSize)
+                _ = bufferOut.writeWithUnsafeMutableBytes(minimumWritableBytes: 0) { ptr in
+                    ptr.copyMemory(from: UnsafeRawBufferPointer(start: dstBuffer.baseAddress, count: Int(messageSize)))
+                    return Int(messageSize)
+                }
+                context.fireChannelRead(wrapInboundOut(bufferOut))
+                return
+            }
+        }
+        logger.info("invalid compressed message received from client @ \(context.channel.remoteAddressDescription), close connection.")
+        context.close(promise: nil)
+    }
+}
+
+final class ChannelCompressionOutboundHandler: ChannelOutboundHandler {
+    typealias OutboundIn = ByteBuffer
+    typealias OutboundOut = ByteBuffer
+
+    private var bufferManager = BufferManager()
+    private var lz4Stream = LZ4_stream_t()
+
+    init() {
+        LZ4_initStream(&lz4Stream, MemoryLayout<LZ4_stream_t>.size);
+    }
+
+    deinit {
+        bufferManager.reset()
```

Comment:
Call `LZ4_freeStream()` ?

Reply:
`LZ4_freeStream()` is a symmetric function for `LZ4_createStream()`, but we use `LZ4_initSteam()` to initialize `LZ4_stream_t` allocated as instance variable and does not suppose any symmetric call to deinitialize. There is `LZ4_resetStream()` which we could use but according documentation the main purpose for that function is to reuse `LZ4_stream_t` object for new stream.